### PR TITLE
Add optional RTS + CTS support to UART

### PR DESCRIPTION
**Pull request by sandeepmistry**

### CODE REVIEW

File: cores/arduino/Uart.cpp [hunk 1]
```diff
@@ -102,11 +129,30 @@ int Uart::peek()
...
+      digitalWrite(uc_pinRTS, LOW);
+    }
+  }
+
+  return c;
 }
 
 size_t Uart::write(const uint8_t data)
 {
+  if (uc_pinRTS != NO_RTS_PIN) {
```
Comment (cmaglie) [inline cores/arduino/Uart.cpp]: why? should RTS influence only Rx?

File: cores/arduino/Uart.cpp [hunk 2]
```diff
@@ -64,6 +84,13 @@ void Uart::IrqHandler()
 {
   if (sercom->availableDataUART()) {
     rxBuffer.store_char(sercom->readDataUART());
+
+    if (uc_pinRTS != NO_RTS_PIN) {
+      // if there is NOT enough space in the RX buffer, de-assert RTS
+      if (rxBuffer.availableForStore() < RTS_RX_THRESHOLD) {
+        digitalWrite(uc_pinRTS, HIGH);
```
Comment (cmaglie) [inline cores/arduino/Uart.cpp]: `digitalWrite` may be slow, we should explore direct register access.

File: cores/arduino/Uart.cpp [hunk 3]
```diff
@@ -102,11 +129,30 @@ int Uart::peek()
...
 }
 
 size_t Uart::write(const uint8_t data)
 {
+  if (uc_pinRTS != NO_RTS_PIN) {
+    // assert RTS
+    digitalWrite(uc_pinRTS, LOW);
+  }
+
+  if (uc_pinCTS != NO_CTS_PIN) {
```
Comment (cmaglie) [inline cores/arduino/Uart.cpp]: mmmhhh there is no other way here? Ideally the CTS pin should mask the tx ISR, so you can always write in the buffer without blocking (until the buffer is full). What is not clear to me is if there is a way to trasparently re-enable the tx ISR.

File: cores/arduino/Uart.cpp (hunk 2 above)
Comment (sandeepmistry) [reply cores/arduino/Uart.cpp]: I've pushed a commit to leverage SERCOM h/w capabilities.

File: cores/arduino/Uart.cpp (hunk 1 above)
Comment (sandeepmistry) [reply cores/arduino/Uart.cpp]: Right, I misunderstood the RTS term - corrected!

File: cores/arduino/Uart.cpp (hunk 3 above)
Comment (sandeepmistry) [reply cores/arduino/Uart.cpp]: The latest commit addresses this.

**Comment by sandeepmistry** [comment]
@cmaglie there's one open question, if we expose the `attachRts` and `attachCts` methods to the user, we need to validate the inputs are the right pins.

The other option is to pass then in the constructor.

**Comment by sandeepmistry** [comment]
I've removed the `attachRts` and `attachCts` API's for now, and made RTS and CTS only configurable via the `Uart` constructor.

**Pull request merged**